pkg/flatrpc/conn.go

```diff
@@ -0,0 +1,185 @@
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/stats"
+)
+
+var (
+	statSent = stats.Create("rpc sent", "Outbound RPC traffic",
+		stats.Graph("traffic"), stats.Rate{}, stats.FormatMB)
+	statRecv = stats.Create("rpc recv", "Inbound RPC traffic",
+		stats.Graph("traffic"), stats.Rate{}, stats.FormatMB)
+)
+
+type Serv struct {
+	Addr *net.TCPAddr
+	ln   net.Listener
+}
+
+func ListenAndServe(addr string, handler func(*Conn)) (*Serv, error) {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		for {
+			conn, err := ln.Accept()
+			if err != nil {
+				if strings.Contains(err.Error(), "use of closed network connection") {
+					break
+				}
+				var netErr *net.OpError
+				if errors.As(err, &netErr) && !netErr.Temporary() {
+					log.Fatalf("flatrpc: failed to accept: %v", err)
+				}
+				log.Logf(0, "flatrpc: failed to accept: %v", err)
+				continue
+			}
+			go func() {
+				c := newConn(conn)
+				defer c.Close()
+				handler(c)
+			}()
+		}
+	}()
+	return &Serv{
+		Addr: ln.Addr().(*net.TCPAddr),
+		ln:   ln,
+	}, nil
+}
+
+func (s *Serv) Close() error {
+	return s.ln.Close()
+}
+
+type Conn struct {
+	conn net.Conn
+
+	sendMu  sync.Mutex
+	builder *flatbuffers.Builder
+
+	data    []byte
+	hasData int
+	lastMsg int
+}
+
+func Dial(addr string, timeScale time.Duration) (*Conn, error) {
+	var conn net.Conn
+	var err error
+	if addr == "stdin" {
+		// This is used by vm/gvisor which passes us a unix socket connection in stdin.
+		conn, err = net.FileConn(os.Stdin)
+	} else {
+		conn, err = net.DialTimeout("tcp", addr, time.Minute*timeScale)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return newConn(conn), nil
+}
+
+func newConn(conn net.Conn) *Conn {
+	return &Conn{
+		conn:    conn,
+		builder: flatbuffers.NewBuilder(0),
+	}
+}
+
+func (c *Conn) Close() error {
+	return c.conn.Close()
+}
+
+type sendMsg interface {
+	Pack(*flatbuffers.Builder) flatbuffers.UOffsetT
+}
+
+// Send sends an RPC message.
+// The type T is supposed to be an "object API" type ending with T (e.g. ConnectRequestT).
+// Sending can be done from multiple goroutines concurrently.
+func Send[T sendMsg](c *Conn, msg T) error {
+	c.sendMu.Lock()
+	defer c.sendMu.Unlock()
+	off := msg.Pack(c.builder)
+	c.builder.FinishSizePrefixed(off)
+	data := c.builder.FinishedBytes()
+	_, err := c.conn.Write(data)
+	c.builder.Reset()
+	statSent.Add(len(data))
+	if err != nil {
+		return fmt.Errorf("failed to send %T: %w", msg, err)
+	}
+	return nil
+}
+
+// Recv received an RPC message.
+// The type T is supposed to be a normal flatbuffers type (not ending with T, e.g. ConnectRequest).
+// Receiving should be done from a single goroutine, the received message is valid
+// only until the next Recv call (messages share the same underlying receive buffer).
+func Recv[T any, PT interface {
```

Comment:
So we must call it already knowing what message type we expect to receive. Is it really the case in our architecture where the entities exchange objects of different types? And I'm not sure we want to keep a separate connection for every kind of the message?

I think it would be closer to the reality if `Recv` internally determined the type of the message and created a pointer of the corresponding type. I'm not sure `flatbuffers` already have this functionality or offload its implementation to the application developers.

Reply:
Ah, okay, it will work fine then.